File: matlab_reference/legacy_python/get_REFSEIS_theme.py
import numpy as np
import pandas as pd
import scipy.io as sio
from matio import load_from_mat
def get_REFSEIS_theme(UTM_X, UTM_Y, terrain, complexity, reg, include_peatlands, Npreq, NPL, Nlay):
    # Load data based on region
    region_files = {
        'Jylland': 'Jylland_seismik_til_usikkerhed.mat',
        'Fyn': None,  # No file specified for Fyn in MATLAB code
        'Sjælland': 'Sjaelland_seismik_til_usikkerhed.mat'
    }
    filename = region_files.get(reg)
    if not filename:
        raise ValueError("Invalid region or missing file for Fyn")

    in_data = sio.loadmat(f'data/{filename}')
    Gmod = load_from_mat(f'data/{filename}')
    Gmod = Gmod['Gmod_pos']

    # Rearrange data
    SEIS = {
        'InfoTable': Gmod,
        'Model_Depths': in_data['dkm_d_pos'],
        'Model_Year': in_data['model_year_pos'].flatten(),
        'DepthInterfaceAboveModel': in_data['dkm_odvlayertop_pos'],
        'ThickGeophysModel': in_data['dkm_odvthk_pos']
    }

    layvec = np.where(SEIS['ThickGeophysModel'][0, :Nlay - NPL] > 1000)[0]
    coords = SEIS['InfoTable'][['xutm_euref89_utm32', 'yutm_euref89_utm32']].values
    _, idx = np.unique(coords, axis=0, return_index=True)
    logind_dupe = np.zeros(len(coords), dtype=bool)
    logind_dupe[idx] = True

    model_year = SEIS['Model_Year']
    geophys_year = SEIS['InfoTable']['d_year'].values
    logind_date = geophys_year <= model_year
    logind = logind_dupe & logind_date

    for key in ['InfoTable', 'Model_Depths', 'Model_Year', 'DepthInterfaceAboveModel', 'ThickGeophysModel']:
        SEIS[key] = SEIS[key][logind]

    nx, ny = len(UTM_X), len(UTM_Y)
    maxl = Nlay - NPL
    Grid = np.zeros((ny, nx, maxl))

    xs = SEIS['InfoTable']['xutm_euref89_utm32'].values
    ys = SEIS['InfoTable']['yutm_euref89_utm32'].values

    local_dist = np.full((ny, nx, maxl), np.nan)
    local_model_thick = np.full((ny, nx, maxl), np.nan)
    complexities = np.full((ny, nx, maxl), np.nan)

    SearchRadius = 8
    comp2range = [100, 500, 400, 250, 100]
    comp2rangePreq = [500] * 5

    # Certainty function for REFSEIS
    def cert_fun(dist, range_, width, sill):
        return np.where(dist <= width, sill, sill * np.exp(-3 * (dist - width)**2 / range_**2))

    for s in layvec:
        MThicks = SEIS['ThickGeophysModel'][:, s]
        complexities[:, :, s] = complexity

        for i in range(ny):
            yv = UTM_Y[i]
            ymin = UTM_Y[max(0, i - SearchRadius)]
            ymax = UTM_Y[min(ny - 1, i + SearchRadius)]
            filt1 = (ys < ymax) & (ys > ymin)

            for j in range(nx):
                xv = UTM_X[j]
                xmin = UTM_X[max(0, j - SearchRadius)]
                xmax = UTM_X[min(nx - 1, j + SearchRadius)]
                filt2 = (xs < xmax) & (xs > xmin)
                filt = filt1 & filt2

                if np.any(filt):
                    
                    local_xs = xs[filt]
                    local_ys = ys[filt]
                    local_thicks = MThicks[filt]

                    distances = np.sqrt((xv - local_xs)**2 + (yv - local_ys)**2)
                    dist = np.min(distances)
                    filt_mindist = distances == dist

                    local_dist[i, j, s] = dist
                    local_model_thick[i, j, s] = np.max(local_thicks[filt_mindist])

    complexity2 = np.where(complexities == 0, np.nan, complexities)
    rangemap = np.copy(complexity2)
    rangemap2 = np.copy(complexity2)

    for i in range(1, 5):
        rangemap[complexity2 == i] = comp2range[i]
        rangemap2[complexity2 == i] = comp2rangePreq[i]

    rangemap[:, :, Npreq - NPL:] = rangemap2[:, :, Npreq - NPL:]
    var0map = np.ones_like(local_dist) * (0.5 * 15)**2
    kernelmap = np.full_like(local_dist, 1)

    certmap = cert_fun(local_dist, rangemap, kernelmap, 1.0 / var0map)
    Grid = 1.0 / certmap
    Grid[np.isnan(Grid)] = 100000

    toi_filter = local_model_thick < 100
    Grid[toi_filter] = 100000

    if include_peatlands:
        New_Grid = np.full((Grid.shape[0], Grid.shape[1], Grid.shape[2] + NPL), 1E4)
        New_Grid[:, :, NPL:] = Grid
        Grid = New_Grid

    s_cords = {'xs': xs, 'ys': ys}
    return Grid, s_cords

File: matlab_reference/N-ret-mod---Source-main/get_REFSEIS_theme.m
function [Grid,s_cords] = get_REFSEIS_theme(UTM_X,UTM_Y,terrain,complexity,reg,include_peatlands,Npreq,NPL,Nlay)

%Load File based on region
switch reg
    case 'Jylland'
        filename = 'jylland_seismik_til_usikkerhed';
    case 'Fyn'
%         filename = 'fyn_seismik_til_usikkerhed';
    case 'Sjælland'
        filename = 'Sjaelland_seismik_til_usikkerhed.mat';
        
end

local_dir = pwd;
in = load([local_dir,'\Geofysik\',filename]);

%Rearrange 'in' struct to 'MEP' struct
switch reg
    case 'Jylland'
        SEIS.InfoTable = in.Gmod_pos;
    case 'Fyn'
        SEIS.InfoTable = in.Gmod_pos;
    case 'Sjælland'
        SEIS.InfoTable = in.Gmod_pos;
end

SEIS.Model_Depths = in.dkm_d_pos;
SEIS.Model_Year = in.model_year_pos;
SEIS.DepthInterfaceAboveModel = in.dkm_odvlayertop_pos;
SEIS.ThickGeophysModel = in.dkm_odvthk_pos;
layvec = find(SEIS.ThickGeophysModel(1,1:(Nlay-NPL))>1000);

Npoints = numel(SEIS.Model_Year);

%Determine duplicate points and calculate logical vector-index to ignore them
[~,id,~] = unique(table2array(SEIS.InfoTable(:,["xutm_euref89_utm32","yutm_euref89_utm32"])),'rows');
logind = zeros(1,Npoints);
logind(id) = 1;

logind_dupe = logical(logind)';

%Determine if the geophysics is younger than the model. If so, filter it
%out.
ModelYear = SEIS.Model_Year;
GeophysYear = table2array(SEIS.InfoTable(:,"d_year"));

logind_date = GeophysYear <= ModelYear;

%Combine Logical Indices
logind = logical(logind_date.*logind_dupe);

%Update the struct
SEIS.InfoTable = SEIS.InfoTable(logind,:);
SEIS.Model_Depths = SEIS.Model_Depths(logind,:);
SEIS.Model_Year = SEIS.Model_Year(logind);
SEIS.DepthInterfaceAboveModel = SEIS.DepthInterfaceAboveModel(logind,:);
SEIS.ThickGeophysModel = SEIS.ThickGeophysModel(logind,:);

% 1) Determine grid size
nx = max(size(UTM_X));
ny = max(size(UTM_Y));
maxl = numel(SEIS.Model_Depths(1,:));

% % 2) Produce grid.
% Grid = zeros(ny,nx,maxl);

% 3) Jeg definerer effective range for udbredelsen af certainty. Første
% element er 50m og bruges når kompleksiteten er ukendt. de næste 4
% elementer svarer til kompleksitet 1-4, hvor 1 er lav og 4 er høj.

comp2range = [100 500 400 250 100];
comp2rangePreq = [500 500 500 500 500];
NP = sum(logind);

if NP > 0

%     cert_fun = @(dist,range,width,sill) (dist<=width).*sill+(dist>width).*sill.*power((19.*power((dist-width)./range,2)+1),-1);
    cert_fun = @(dist,range,width,sill) (dist<=width).*sill+(dist>width).*sill.*(exp(-(3*power((dist-width),2))./(power(range,2))));
    % 6) Jeg skal besøge hver celle og lægge bidrag for de nærmeste boringer
    SearchRadius = 8; %in units of cells

    LLx = min(UTM_X);
    LLy = min(UTM_Y);

    xs = table2array(SEIS.InfoTable(:,"xutm_euref89_utm32"));
    ys = table2array(SEIS.InfoTable(:,"yutm_euref89_utm32"));
    maxl = Nlay-NPL;
    local_dist = NaN([size(terrain),maxl]);
    complexities = NaN([size(terrain),maxl]);

    for s = layvec

        MThicks = SEIS.ThickGeophysModel(:,s);
        complexities(:,:,s) = complexity;

        parfor i = 1:ny
            Ymin = max([1 i-SearchRadius]);
            Ymax = min([ny i+SearchRadius]);
            
            ymin = UTM_Y(Ymin);
            ymax = UTM_Y(Ymax);
            
            yv = UTM_Y(i);

            filt1 = ys<ymax & ys>ymin;

            for j = 1:nx

                Xmin = max([1 j-SearchRadius]);
                Xmax = min([nx j+SearchRadius]);

                xmin = UTM_X(Xmin);
                xmax = UTM_X(Xmax);

                xv = UTM_X(j);


                filt2 = xs<xmax & xs>xmin;
                filt = filt1 & filt2;

                npoints = sum(filt);

                if npoints > 0
                    local_xs = xs(filt);
                    local_ys = ys(filt);
                    local_thicks = MThicks(filt)';

                    distances = sqrt(power(xv-local_xs,2)+power(yv-local_ys,2));
                    dist = min(distances);

                    filt_mindist = distances == dist;

                    local_dist(i,j,s) = dist;
                    local_model_thick(i,j,s) = max(local_thicks(filt_mindist));
                end
            end
        end
   end
end

complexity2 = complexities;
complexity2(complexity==0) = NaN;
rangemap = complexity2;
rangemap2 = complexity2;

%Calculate var0 map
var0map = ones(size(local_dist)).*power(0.5*15,2);

rangemap(~isnan(complexity2)) = comp2range(complexity2(~isnan(complexity2))+1);
rangemap2(~isnan(complexity2)) = comp2rangePreq(complexity2(~isnan(complexity2))+1);
rangemap(:,:,Npreq-NPL:end) = rangemap2(:,:,Npreq-NPL:end);

kernelmap = local_dist*0+1;
certmap = cert_fun(local_dist,rangemap,kernelmap,1./var0map);

Grid = 1./certmap;
Grid(isnan(Grid)) = 100000;

toi_filter = local_model_thick < 100;
Grid(toi_filter) = 100000;        

if include_peatlands == 1
    New_Grid = 1E4*ones(size(Grid,1),size(Grid,2),size(Grid,3)+NPL);
    New_Grid(:,:,(NPL+1):end) = Grid;
    Grid = New_Grid;
end

s_cords.xs = xs;
s_cords.ys = ys;

end